include/igl/opengl/ViewerData.cpp

```diff
@@ -795,4 +797,101 @@ IGL_INLINE void igl::opengl::ViewerData::updateGL(
       meshgl.points_F_vbo(i) = i;
     }
   }
+
+  if (meshgl.dirty & MeshGL::DIRTY_VID_LABELS)
+  {
+    std::string vertCount = std::to_string(V.rows());
+    int totalVertCharacters = 0;
+    int i=1;
+    while(i<vertCount.length())
+    {
+      totalVertCharacters += (i)*9*pow(10,i-1); i++;
```

Comment:
Why do you need this?

Reply:
I wanted a more efficient way to determine the number of characters in all vertex labels. 
I was trying to find a faster forloop instead of iterating over every single vertex and summing up the string lengths since we know the contents of all the stringe beforehand. I sketched this out on paper and it seemed to work correctly and it's O(log(v)) instead of O(v).

I condensed it to 
```
    int totalVertCharacters = 0;
    for(int i=1; i<vertCount.length(); i++)
      totalVertCharacters += (i)*9*pow(10,i-1);
    totalVertCharacters += (vertCount.length() * (V.rows() - pow(10,vertCount.length()-1))) + 1;
```
and just did a google search and seems a similar algorithm is out there: https://www.geeksforgeeks.org/count-total-number-digits-1-n/ except we shift everything by 1 (0 to n-1). It's a lot cleaner than mine.

Adapting my code to this form:
```
    int totalVertCharacters = 1; // 0th Index
    for(int i=1; i<V.rows(); i*=10)
      totalVertCharacters += (V.rows() - i);
```

Does this sound good to keep? The final versions would be 
```
// Vertices
    int totalVertCharacters = 1; // 0th Index
    for(int i=1; i<V.rows(); i*=10)
    {
      totalVertCharacters += (V.rows() - i);
    }
// Faces
    int totalFaceCharacters = 1; // 0th Index
    for(int i=1; i<F.rows(); i*=10)
    {
      totalFaceCharacters += (F.rows() - i);
    }

```

Reply:
Hmm that seems like an unnecessary complication. You shouldn't be worried about speed, since you're making the same loop later on anyway (from 1 to #V, then for each char in the vertex label).

For simplicity I would produce suggest producing 1 std::vector<std::string> + 1 matrix of label positions for both the vertex ids and the face ids, and reuse the code for the general "labels" to fill the VBOs for vertex/face labels (--> less code duplication).

Reply:
This is a great idea, will do so now thanks :-)

Reply:
Though this being said, that would mean we would need to initialize those data structures in addition to assigning them to the VBOs. Also, taking a closer look though it's ideal to not have so much duplicated code I'm having a hard time seeing how to avoid having 3 copies of the VBO-populating forloops can be replaced with one. Did you mean this as well? If we were to do so I imagine it being a switch case as to which buffer needs to refreshed then dynamically populating just that one, though that restricts us to updating all three buffers in one instance of `updateGL()`

Reply:
You could encapsulate all the VBO machinery for the labels into a separate struct, e.g. call it `TextGL`. Then MeshGL would have 3 members variables:
```cpp
TextGL vertex_labels;
TextGL face_labels;
TextGL custom_labels;
```

That would simplify a lot of code I think. You also need to be careful about the `bind_font_atlas()` function: `igl::png::texture_from_png` is generating the PNG and generating a new texture on the GPU. Right now you're generating a new texture everytime the label data is updated, so you need to store the handle in the TextGL class, and call `glDeleteTextures()` to destroy it. Although, since it's gonna be a texture shared by all meshes, we should probably make it static and use Meyer's singleton pattern to initialize it on the first access (and thus never destroy it... unless we use some global ref counting which may be a bit overkill).

Reply:
I've encapsulated the labels into structs as well as abstracted the functions they are populated in and it reduced a lot of code so that's good :-)

But I'm definitely noticing the effects of `bind_font_atlas`. It's being corruptive when used in conjunction with the overlay lines and points. I'm looking at how other textures are updated in libigl to try to mimic that...

Reply:
Alright, I resolved the aforementioned issue. 
WRT `glDeleteTextures()`, what about storing the font atlas handle in MeshGL.h, like the other texture handle that's there, initialize the font atlas in `MeshGL::init()`, then just bind it each time we render any of the text labels? I've just pushed code which does this, notably `MeshGL::init_font_atlas()` and `MeshGL::bind_font_atlas()`

Reply:
Yes that looks better indeed. For the `init_font_atlas` I guess we can keep it as is for now, that shouldn't be an issue.